scripts/ci/notification/matrix_notification.sh

```diff
@@ -0,0 +1,87 @@
+#!/bin/bash
+
+set -euo pipefail
+shopt -s inherit_errexit
+
+SCRIPT_DIR="$(dirname "$0")"
+
+function fail() {
+    printf '%s\n' "$1" >&2
+    exit 1
+}
+
+function notify() {
+    [[ -z "$1" ]] && fail "Event type not provided."
+    [[ "$1" != "failure" && "$1" != "success" && "$1" != "release" ]] && fail "Wrong event type."
+    local event="$1"
+    local formatted_message
+
+    # FIXME: Checking $CIRCLE_PULL_REQUEST would be better than hard-coding branch names
+    # but it's broken. CircleCI associates runs on develop/breaking with random old PRs.
+    [[ "$BRANCH" == "develop" || "$BRANCH" == "breaking" ]] || { echo "Running on a PR or a feature branch - notification skipped."; exit 0; }
+
+    # The release notification only makes sense on tagged commits. If the commit is untagged, just bail out.
+    [[ "$event" == "release" ]] && { [[ $TAG != "" ]] || { echo "Not a tagged commit - notification skipped."; exit 0; } }
+
+    formatted_message="$(format_predefined_message "$event")"
+
+    curl "https://${MATRIX_SERVER}/_matrix/client/v3/rooms/${MATRIX_NOTIFY_ROOM_ID}/send/m.room.message" \
+        --request POST \
+        --include \
+        --header "Content-Type: application/json" \
+        --header "Accept: application/json" \
+        --header "Authorization: Bearer ${MATRIX_ACCESS_TOKEN}" \
+        --data "$formatted_message"
+}
+
+function circleci_workflow_name() {
+    # Workflow name is not exposed as an env variable. Has to be queried from the API.
+    # The name is not critical so if anything fails, use the raw workflow ID as a fallback.
+    local workflow_info workflow_name
+    workflow_info=$(curl --silent "https://circleci.com/api/v2/workflow/${CIRCLE_WORKFLOW_ID}") || true
+    workflow_name=$(echo "$workflow_info" | grep -E '"\s*name"\s*:\s*".*"' | cut -d \" -f 4 || echo "$CIRCLE_WORKFLOW_ID")
+    echo "$workflow_name"
+}
+
+function circleci_job_name() {
+    local job
+    [[ $CIRCLE_NODE_TOTAL == 1 ]] && job="${CIRCLE_JOB}"
+    [[ $CIRCLE_NODE_TOTAL != 1 ]] && job="${CIRCLE_JOB} (run $((CIRCLE_NODE_INDEX + 1))/${CIRCLE_NODE_TOTAL})"
+    echo "$job"
```

Comment:
```suggestion
    [[ $CIRCLE_NODE_TOTAL == 1 ]] && echo "${CIRCLE_JOB}"
    [[ $CIRCLE_NODE_TOTAL != 1 ]] && echo "${CIRCLE_JOB} (run $((CIRCLE_NODE_INDEX + 1))/${CIRCLE_NODE_TOTAL})"
```

Reply:
I'm not sure about this change. Only echoing in the first condition will not return from the function even when `CIRCLE_NODE_TOTAL == 1`, thus the script will continue to the second condition, which will evaluate to `false` in this case, and the exit status of the function will then be 1 instead of 0. As we `set -e`, this will terminate the script. We should keep it as it is or do:

```sh
(( $CIRCLE_NODE_TOTAL == 1 )) && echo "${CIRCLE_JOB}" && return
(( $CIRCLE_NODE_TOTAL != 1 )) && echo "${CIRCLE_JOB} (run $((CIRCLE_NODE_INDEX + 1))/${CIRCLE_NODE_TOTAL})"
```

The return will ensure that whatever is evaluated in the first condition is returned as exit status of the function.

Reply:
I added the return. Let me know if you prefer the other way ;)

Reply:
Ah, right. I missed the fact that these conditions are not entirely equivalent to `if`/`else` and you can run into corner cases like this. `return` looks like a good way out.